File: app/Sources/TarmacKit/CardChrome.swift
/// Pure rule for a card's resting visual chrome — its border role and whether
/// the resize handles show. Collapses `focused`/`selected` into one "active
/// card" state (the teal ring + handles, always together) and keeps `prime`
/// (the keyboard target) out of the border entirely: prime is signalled by
/// header tint + shadow in the AppKit layer, never by a border here. Kept in
/// TarmacKit so the priority and the "ring ⟺ handles" invariant are unit-tested
/// away from AppKit (mirrors `EscFocusAction` / `FocusedClose`).
public enum CardChrome {
    /// Visual-state inputs for one card.
    public struct State: Equatable {
        public var dead: Bool
        public var detached: Bool
        public var fresh: Bool
        /// The keyboard target — intentionally NOT a border input.
        public var prime: Bool
        public var focused: Bool
        public var selected: Bool

        public init(
            dead: Bool = false,
            detached: Bool = false,
            fresh: Bool = false,
            prime: Bool = false,
            focused: Bool = false,
            selected: Bool = false
        ) {
            self.dead = dead
            self.detached = detached
            self.fresh = fresh
            self.prime = prime
            self.focused = focused
            self.selected = selected
        }
    }

    /// The resting border role; `CardView` maps each case to a `Theme` colour.
    public enum BorderRole: Equatable {
        /// Dead or detached — muted line.
        case muted
        /// The agent-opened "fresh" halo.
        case agent
        /// The unified active ring (teal) — an active card that is not dead/detached.
        case focus
        /// Nothing notable — the plain line.
        case plain
    }

    /// True when the card is the user's active target — a single click
    /// (`focused`) or an explicit header/handle grab (`selected`). NOT
    /// suppressed by `dead`/`detached`: a dead card stays resizable via a
    /// header grab, so its handles can still show even though its border is muted.
    public static func showsHandles(_ s: State) -> Bool {
        s.focused || s.selected
    }

    /// The resting border role, highest priority first:
    ///   dead || detached    -> .muted   (handles may still show — resize)
    ///   active (focus/sel.)  -> .focus   (the unified ring)
    ///   fresh                -> .agent
    ///   else                 -> .plain
    /// `prime` never appears — the keyboard target is signalled outside the border.
    public static func borderRole(_ s: State) -> BorderRole {
        if s.dead || s.detached { return .muted }
        if showsHandles(s) { return .focus }
        if s.fresh { return .agent }
        return .plain
    }
}

File: desktop/src/kit/cardChrome.ts
// Port of TarmacKit/CardChrome.swift — the pure rule for a card's resting
// visual chrome: its border role and the set of resize handles. Collapses
// `focused`/`selected` into one "active card" state for the teal ring; keeps
// both `prime` (keyboard target) and `fresh` (agent-opened, unread) OUT of the
// border — prime is signalled by header tint + shadow, fresh by its halo + "✚ now"
// meta in the view layer, never by a border here.
// Handle hit-areas are always live (hover-revealed via CSS); borderRole stays
// focus-gated so idle cards never show the focus ring.

import type { Handle } from "./resize";

export interface CardChromeState {
  dead: boolean;
  detached: boolean;
  /** Agent-opened and unread — signalled by halo + "✚ now" meta, NOT a border input. */
  fresh: boolean;
  /** The keyboard target — NOT a border input. */
  prime: boolean;
  focused: boolean;
  selected: boolean;
}

/** The resting border role; the card view maps each to a theme colour. */
export type BorderRole = "muted" | "focus" | "plain";

export const cardChromeState = (s: Partial<CardChromeState> = {}): CardChromeState => ({
  dead: false,
  detached: false,
  fresh: false,
  prime: false,
  focused: false,
  selected: false,
  ...s,
});

/**
 * True when the card is the user's active target — used by borderRole to gate
 * the focus ring. NOT used to gate handle hit-areas (those are always live).
 */
export function showsHandles(s: CardChromeState): boolean {
  return s.focused || s.selected;
}

const ALL_HANDLES: readonly Handle[] = ["tl", "t", "tr", "r", "br", "b", "bl", "l"];

/**
 * The ordered set of resize handle ids for a card. `hasClose` drops `"tr"` so
 * the close button and the corner handle never collide; the `"t"` edge strip is
 * right-trimmed in CSS via the `has-close` class.
 */
export function cardHandles(hasClose: boolean): Handle[] {
  return hasClose ? ALL_HANDLES.filter(h => h !== "tr") : [...ALL_HANDLES];
}

/**
 * The resting border role, highest priority first:
 *   dead || detached   -> "muted"  (handles may still show — resize)
 *   active (focus/sel.) -> "focus"  (the unified ring)
 *   else               -> "plain"
 * Neither `prime` nor `fresh` appears — both are signalled outside the border.
 */
export function borderRole(s: CardChromeState): BorderRole {
  if (s.dead || s.detached) return "muted";
  if (showsHandles(s)) return "focus";
  return "plain";
}
